src/nunavut/lang/cpp/support/serialization.hpp

```diff
@@ -1,73 +1,318 @@
- * @param  length_bits      The total length of bits to copy. The caller must ensure
- *                          that the size of src and dst are >= this value.
- *
- * @return The number of bits copied.
- */
-template <typename SizeType, typename ByteType>
-SizeType copyBitsAlignedToUnaligned(const ByteType* const  src,
-                                    std::vector<ByteType>& dst,
-                                    const SizeType         dst_offset_bits,
-                                    const SizeType         length_bits)
+template <unsigned BitLen>
+struct NativeFloatSelector
 {
-    if (nullptr == src || length_bits == 0)
-    {
-        return 0;
-    }
-    constexpr SizeType bits_in_SizeType = sizeof(ByteType) * 8U;
-    SizeType           bits_copied      = 0;
-    SizeType           offset_bits      = dst_offset_bits;
-    const SizeType     local_offset     = dst_offset_bits % bits_in_SizeType;
-    do
-    {
-        SizeType       current_byte       = offset_bits / bits_in_SizeType;
-        const SizeType bits_from_src_byte = bits_in_SizeType - local_offset;
-        dst[current_byte] &= static_cast<ByteType>(0xFF >> bits_from_src_byte);
-        dst[current_byte] |= static_cast<ByteType>(src[current_byte] << local_offset);
-        offset_bits += bits_in_SizeType;
-        bits_copied += std::min(length_bits, bits_from_src_byte);
-        if (offset_bits < length_bits)
-        {
-            dst[current_byte] |= static_cast<ByteType>(src[offset_bits / bits_in_SizeType] >> bits_from_src_byte);
-            bits_copied += local_offset;
-        }
-        else
+    struct ErrorNoSuchFloat;
+    typedef typename std::conditional<
+        sizeof(float) * 8 >= BitLen,
+        float,
+        typename std::conditional<
+            sizeof(double) * 8 >= BitLen,
+            double,
+            typename std::conditional<sizeof(long double) * 8 >= BitLen, long double, ErrorNoSuchFloat>::type>::type>::
+        type float_type;
+
+    typedef typename std::conditional<
+        BitLen <= 16,
+        std::uint16_t,
+        typename std::conditional<BitLen <= 32,
+                                  std::uint32_t,
+                                  typename std::conditional<BitLen <= 64, long double, ErrorNoSuchFloat>::type>::type>::
+        type int_storage_type;
+};
+
+namespace IEEE754Conversion
+{
+namespace internal
+{
+template <unsigned BitLen>
+void enforceIeee()
+{
+    /*
+     * Some compilers may have is_iec559 to be defined false despite the fact that IEEE754 is supported.
+     * An acceptable workaround would be to put an #ifdef here.
+     */
+    static_assert(std::numeric_limits<typename NativeFloatSelector<BitLen>::Type>::is_iec559,
+                  "We were not able to find a compatible floating-point type for this compiler.");
+}
+
+union Fp32
+{
+    std::uint32_t u;
+    float         f;
+};
+
+}  // namespace internal
+
+template <typename NativeType>
+NativeType nativeIeeeToHalf(float value)
```

Comment:
```suggestion
NativeType nativeIeeeToHalf(const float value)
```

Reply:
May I suggest that we introduce macros for the language features that are not available in older versions of C++ to make better use of the newer versions? Particularly, I would like to use `[[nodiscard]]` via some `NUNAVUT_NODISCARD` which evaluates to nothing in C++11/14.

Reply:
Going to delete all this for now.